sdk/inc/azure/iot/az_iot_pnp_client.h

```diff
@@ -0,0 +1,730 @@
+} az_iot_pnp_client;
+
+/**
+ * @brief Gets the default Azure IoT PnP Client options.
+ * @details Call this to obtain an initialized #az_iot_pnp_client_options structure that can be
+ *          afterwards modified and passed to az_iot_pnp_client_init().
+ *
+ * @return #az_iot_pnp_client_options.
+ */
+AZ_NODISCARD az_iot_pnp_client_options az_iot_pnp_client_options_default();
+
+/**
+ * @brief Initializes an Azure IoT PnP Client.
+ *
+ * @param[out] out_client The #az_iot_pnp_client to use for this call.
+ * @param[in] iot_hub_hostname The IoT Hub Hostname.
+ * @param[in] device_id The Device ID.
+ * @param[in] model_id The root DTDL interface of the #az_iot_pnp_client.
+ * @param[in] options A reference to an #az_iot_pnp_client_options structure. Can be `NULL`.
+ *
+ * @pre \p out_client must not be `NULL`.
+ * @pre \p iot_hub_hostname must be a valid, non-empty #az_span.
+ * @pre \p device_id must be a valid, non-empty #az_span.
+ * @pre \p model_id must be a valid, non-empty #az_span.
+ *
+ * @return An #az_result value indicating the result of the operation.
+ */
+AZ_NODISCARD az_result az_iot_pnp_client_init(
+    az_iot_pnp_client* out_client,
+    az_span iot_hub_hostname,
+    az_span device_id,
+    az_span model_id,
+    az_iot_pnp_client_options const* options);
+
+/**
+ * @brief The HTTP URI Path necessary when connecting to IoT Hub using WebSockets.
+ */
+#define AZ_IOT_PNP_CLIENT_WEB_SOCKET_PATH "/$iothub/websocket"
+
+/**
+ * @brief The HTTP URI Path necessary when connecting to IoT Hub using WebSockets without an X509
+ * client certificate.
+ * @note Most devices should use #AZ_IOT_PNP_CLIENT_WEB_SOCKET_PATH. This option is available for
+ * devices not using X509 client certificates that fail to connect to IoT Hub.
+ */
+#define AZ_IOT_PNP_CLIENT_WEB_SOCKET_PATH_NO_X509_CLIENT_CERT \
+  AZ_IOT_PNP_CLIENT_WEB_SOCKET_PATH "?iothub-no-client-cert=true"
+
+/**
+ * @brief Gets the MQTT user name.
+ *
+ * The user name will be of the following format:
+ *
+ * `{iothubhostname}/{device_id}/?api-version=2020-09-30&{user_agent}&model-id={model_id}`
+ *
+ * @param[in] client The #az_iot_pnp_client to use for this call.
+ * @param[out] mqtt_user_name A buffer with sufficient capacity to hold the MQTT user name.
+ *                            If successful, contains a null-terminated string with the user name
+ *                            that needs to be passed to the MQTT client.
+ * @param[in] mqtt_user_name_size The size, in bytes of \p mqtt_user_name.
+ * @param[out] out_mqtt_user_name_length __[nullable]__ Contains the string length, in bytes, of
+ *                                                      \p mqtt_user_name. Can be `NULL`.
+ *
+ * @pre \p client must not be `NULL`.
+ * @pre \p mqtt_user_name must not be `NULL`.
+ * @pre \p mqtt_user_name_size must be greater than 0.
+ *
+ * @return An #az_result value indicating the result of the operation.
+ */
+AZ_NODISCARD AZ_INLINE az_result az_iot_pnp_client_get_user_name(
```

Comment:
I think we should avoid inling APIs just because of how it is implemented _today_. If inlining an API makes sense from a performance perspective, or other customer-facing benefit/constraint, let's do it.

Otherwise, it is generally better to not inline implementations into the header files. Our other `_get_user_name` methods are not inlined, so it makes me think this is not something we should inline, here either.

The same applies to all other places where we inline only because the implementation is simple enough to allow it.